### PR TITLE
[service.listenbrainz] 0.3.1

### PR DESCRIPTION
**Pull request by Freso**
### Description

Enable submitting listens to [ListenBrainz](https://listenbrainz.org/).

See also https://github.com/xbmc/repo-scripts/pull/2160 for version for Leia.

### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [script.foo.bar] 1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

File: service.listenbrainz/addon.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		<summary lang="en_US">ListenBrainz listener</summary>
+		<summary lang="sv_SE">ListenBrainz lyssnar</summary>
+		<description lang="da_DK">Lytter for ListenBrainz der sender information om den musik du lytter til i Kodi til ListenBrainz</description>
+		<description lang="en_GB">The ListenBrainz listener will submit info of the music you’re listening to in Kodi to ListenBrainz</description>
+		<description lang="en_NZ">The ListenBrainz listener will submit info of the music you’re listening to in Kodi to ListenBrainz</description>
+		<description lang="en_US">The ListenBrainz listener will submit info of the music you’re listening to in Kodi to ListenBrainz</description>
+		<description lang="sv_SE">Lyssnar för ListenBrainz som skickar information om den musik du lyssnar på i Kodi till ListenBrainz</description>
+		<language></language>
+		<platform>all</platform>
+		<license>GNU GENERAL PUBLIC LICENSE Version 2</license>
```
Comment (basrieter) [inline service.listenbrainz/addon.xml:21]: Please use the license SPDX (See https://spdx.org/licenses/)

File: service.listenbrainz/listener.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+        log('nowplaying', SESSION)
+        result = listenbrainz.submit_playingnow(listen)
+        if not result:
+            return
+        # parse response
+        if 'status' in result and result['status'] == 'ok':
+            return
+        elif 'error' in result:
+            msg = result['error']
+            xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/listener.py:108]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/listener.py [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+            videos=videos,
+            radio=radio)
+        # init the monitor class (re-init when settings have changed)
+        self.monitor = MyMonitor(
+            action=self._get_settings,
+            user=user,
+            token=token,
+            server=server)
+        if not listenbrainz.validate_server():
+            xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/listener.py:70]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/listener.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+            write_file(self.file, self.queue)
+            return
+        # parse response
+        if 'status' in result and result['status'] == 'ok':
+            # remove submitted items from the list
+            self.queue = self.queue[50:]
+        elif 'error' in result:
+            code = result['code']
+            msg = result['error']
+            xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/listener.py:167]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/utils.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+    if puser or ptoken:
+        if puser != user:
+            # username changed
+            pass  # TODO: Validate username
+        if ptoken != token:
+            # token changed
+            listenbrainz.validate_token()
+    elif not (user and token):
+        # no username or token
+        xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/utils.py:50]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/utils.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+                requests.exceptions.ConnectionError,
+                ValueError):
+            pass
+        return False
+
+    def validate_token(self):
+        """Validate a ListenBrainz token."""
+        endpoint = 'validate-token'
+        if not self.token:
+            xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/utils.py:167]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/utils.py [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+                'Notification(%s,%s,%i)' %
+                (LANGUAGE(32011), LANGUAGE(32027), 7000))
+            return False
+        payload = {'token': self.token}
+        response = self._get(endpoint, payload=payload).json()
+        if 'message' in response:
+            token_valid = response['message'] == 'Token valid.'
+            if token_valid is False:
+                xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/utils.py:176]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/utils.py [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+from helpers import is_local, get_url_data
+from mapping import kodi_mapping
+
+__addon__ = xbmcaddon.Addon()
+__addonid__ = __addon__.getAddonInfo('id')
+__version__ = __addon__.getAddonInfo('version')
+
+HEADERS = {'User-Agent': 'Kodi Media center', 'Accept-Charset': 'utf-8'}
+LANGUAGE = __addon__.getLocalizedString
+DATAPATH = xbmc.translatePath(
```
Comment (basrieter) [inline service.listenbrainz/utils.py:20]: use `xbmcvfs.translatePath`, `xbmc.translatePath` is deprecated.

**Review by basrieter — CHANGES_REQUESTED** [review]
Please review my changes.

**Comment by Freso** [comment]
> Please review my changes.

These should all have been addressed in 0.3.1. :)

**Review by basrieter — APPROVED** [review]

**Pull request merged**